app/com/arpnetworking/rollups/MetricsDiscovery.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2019 Dropbox Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.arpnetworking.rollups;
+
+import akka.actor.AbstractActorWithTimers;
+import akka.actor.ActorRef;
+import akka.actor.Status;
+import akka.pattern.PatternsCS;
+import com.arpnetworking.kairos.client.KairosDbClient;
+import com.arpnetworking.kairos.client.models.KairosMetricNamesQueryResponse;
+import com.arpnetworking.play.configuration.ConfigurationHelper;
+import com.arpnetworking.steno.Logger;
+import com.arpnetworking.steno.LoggerFactory;
+import com.typesafe.config.Config;
+import scala.concurrent.duration.FiniteDuration;
+
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Pattern;
+import javax.inject.Inject;
+
+/**
+ * Actor for discovering the list of metrics available to be rolled up on a periodic basis.
+ *
+ * This actor maintains an internal set of metric names and acts as a source for downstream
+ * actors that perform the actual rollups.  This is intended to be used as a singleton in the
+ * cluster.
+ *
+ * @author Gilligan Markham (gmarkham at dropbox dot com)
+ */
+public class MetricsDiscovery extends AbstractActorWithTimers {
+
+    @Override
+    public Receive createReceive() {
+        return receiveBuilder()
+                .matchEquals(FETCH_MSG,
+                        work -> fetchMetricsForRollup())
+                .match(KairosMetricNamesQueryResponse.class,
+                        this::updateMetricsSet)
+                .match(Status.Failure.class,
+                        failure -> LOGGER.warn("Failed to get metrics from Kairo", failure.cause()))
+                .match(MetricFetch.class, work -> {
+                    final Optional<String> metricName = getNextMetric();
+                    if (metricName.isPresent()) {
+                        getSender().tell(metricName.get(), getSelf());
+                    } else {
+                        getSender().tell(NoMoreMetrics.getInstance(), getSelf());
```

Comment:
Should the reply of `NoMoreMetrics` carry the time until the next refresh?

Reply:
I'm not sure if I can really calculate that.  I guess I could stamp the next fetch time as the last_start + interval but since it's using a Akka periodic timer I feel that's just guessing.  I would assume the cost of sending a periodic message to see if anything is available is relatively low and therefore not worth the risk of having some time skew problem causing metrics to be missed for a period.

Reply:
You could always add a configurable jitter to the advertised target. I think any inadvertent polling would be acceptable. This is similar to the deadline walk that Spencer is dealing with.

The one difference is that your actor is remote to the timestamp. So perhaps sending an interval instead of a target date/time would avoid any host-to-host clock skews. I agree this does nothing to account for Akka timer drift, but the jitter should make-up for most/all of that.

Reply:
I don't fully agree that the complexity vs periodic polling is worth it but I added the number of milliseconds to the next refresh in the message.